### ISSUE TITLE
File field suggestion

### ISSUE THREAD

**Issue by ginad** [jonathangeiger/kohana-jelly#61]
suggested feature for file field: 
1. if column is not empty then delete the file first then upload the new file and update the column value.(we can put an option in the initialize function to identify if the file is to be deleted or not)
2. if column is not empty and submitted, then do nothing because right now it puts null instead of retaining the column content.

example code: 

```
/**
 * return string file name if new file uploaded
 */
public function save($model, $value, $loaded)
{
    if($loaded)
    {
        $rec = Jelly::select($this->model, $model->id());
        $name = $this->name;
        $val = $rec->$name; // restore column content
    }
        else{ $val = "";
    }

    // Upload a file?
    if (upload::valid($value))
    {
        if (FALSE !== ($filename = upload::save($value, null, $this->path)))
        {
            $f = pathinfo($filename);
            // return path and file name and not the absolute location of the file
            $val = $this->path . $f['basename'];
        }
    }
    else
    {
        $val = $this->default; 
    }

    return $val;

} // end save
```

**Comment by jonathangeiger** [comment]
I was hoping we could generate a bit of discussion on this. I'm fine with implementing number 1, but I'm not so sure about number 2. If the upload fails for whatever reason, what should we be doing? Reverting to the last value, throwing an exception, setting the default?

Banks, where are you? 

;D

**Comment by banks** [comment]
Busy week(end)...

Not sure I understand point 2 at all. If the file is non empty and submitted isn't that the same case as point 1?

If upload fails, I think the field's validation should somehow catch it so error handling is retained without additional logic in the form/controller. I've never used or lookedat the code for this field though so I'm not sure if that is possible. Perhaps we could create some static helper functions in the field to allow validating the $_FILES data or something as part of the natural validation.

If an error occurs whilst handling files in save() then an exceptions should be thrown. It is hard though because at this point the new data has already been written to the DB. You could change back to the original value (which should still be in memory at this point) and resave before throwing the exception? A kind of faux transaction. Of you could leave this for the developer to handle.

**Comment by ginad** [comment]
oh, sorry what i mean in point 2 is if there is no file to be uploaded(field is empty) then we must not overwrite the current content of the column, because right now it erases the original content and replace it with null instead of retaining the original content.

**Comment by vitch** [comment]
I definitely agree with the need for both of these features. 

Re 1. - just to clean up after yourself and keep the server clean (plus the fact that people might be replacing an uploaded file because the current one has an error which should be hidden).

Re 2. - It is pretty normal to have an form which has a file upload field. If nothing is put in the upload field then the current file is left as is and the field in the DB shouldn't be updated. If a new file is successfully uploaded then it replaces the existing one.

**Comment by jonathangeiger** [comment]
Regarding number 2: If all empty() values don't overwrite the field, how are you supposed to empty it?

**Comment by vitch** [comment]
Good point. From a UI point of view I'd normally accomplish it by having a "delete" checkbox next to the item. Although that isn't always necessary (e.g. in this case I have many related files for a given news item. Each file is a title and a path. If you wanted to delete it you would delete the entire record rather than just deleting the path field).
I'm looking into implementing the functionality on my fork of Jelly - if so I'll make behaviour 2 optional and will provide a method of actually deleting the contents of the field (maybe differentiating between NULL and an empty string?)

**Comment by vitch** [comment]
My implementations of these are available at vitch/kohana-jelly@3356a15cf2577968688c0bd6306e9e974d6d2273 and vitch/kohana-jelly@0533cd7643b131a4eb76a8e6631d081df90563af (on my issue_61 branch: http://github.com/vitch/kohana-jelly/commits/issue_61 ).

Hope it helps :)

**Comment by vitch** [comment]
Whoops - and a bug fix in vitch/kohana-jelly@3216a32b2edaed9906357a1aee6438919c8daf91 :)

**Comment by jonathangeiger** [comment]
Partially closed by 24814b5a16ac3b49470a39494cd52bbc8a81ea08. Field_Files can now delete their old files. This is untested.

Just to add on to this, I'm still not convinced about #2. In my opinion, it's the developer's responsibility to keep track of whether or not a field should be updated. This is how every other field works, so to have Field_File act differently seems unintuitive.

I'll leave this open for more discussion.

**Comment by vitch** [comment]
Would you care to comment on why you went for this approach vs. the approach I took in the code above? I'm still pretty new to Jelly so I'm sure there is a good reason but I'd like to learn from it.

Re. num 2 above, I can see where you are coming from. If no file was uploaded you could just not call $model->set with the file field. However, this means that you would need to duplicate the "check if a file was uploaded before pushing it into the model" logic in each instance where you have a form with a Field_File.

In terms of this being different to how every other field works, the reason for that is because of how [input type="upload"] works in browsers. For a normal text field, if the user loads the form and makes no changes then the same value is submitted. For a file input, if the user makes no changes then the old value isn't (and can't be) automatically propagated.

Not sure what other people think about this?

**Comment by jonathangeiger** [comment]
My approach wasn't entirely different than yours, really. I wanted to avoid the extra query so I added a way in the model to return the original values. I also wanted to make it optional to delete the old file because sometimes it's useful to keep the files around for archiving purposes. The rest was just cleaning up the code a bit.

I do appreciate your fork and I mostly went off that. I feel bad now because I realize that I forgot to include a credit in the commit. How can I make it up to you?

I see your points about #2. I'm kind of torn over it because both Banks and I are adamant about decoupling the ORM layer and the form layer, so Jelly shouldn't necessarily care about how file inputs work, but this field is inextricably tied to their behavior.

Looking at Sprig's Field_Image, it appears that his works the same way. If the value is changed it's overwritten.

**Comment by vitch** [comment]
No problem. My approach started from the example by @ginad above and was just tweaked...

When I've passed my current urgent deadline I'll look into re-integrating your approach and will let you know if there is any problems.

Re number 2, I am coming around to your approach but still not 100% convinced. I'll let you guys decide...

**Comment by jonathangeiger** [comment]
Partially closed by 24814b5a16ac3b49470a39494cd52bbc8a81ea08. Field_Files can now delete their old files. This is untested.

**Comment by vitch** [comment]
In my fork I've integrated these changes and tested and it seems to work well...

I also got rid of my code in the class for deciding whether to overwrite the existing value and replaced it with this in the controller:

```
if ($_FILES[$field_id]['tmp_name'] != '' && $_FILES[$field_id]['size'] != 0) {
    $model->set(array($field_id => Arr::get($_FILES, $field_id)));
}
```

(e.g. I don't call ->set() on the model unless a file was uploaded). I actually think maybe this is cleaner so I'm happy with how it is...
